specs/dwritecore/DWriteCore.md

```diff
@@ -1235,3 +2038,916 @@ DWRITE_BEGIN_INTERFACE(IDWriteFactory8, "202D44F0-84AC-4D6F-B8D0-704B27675CAA")
+    /// glyphAttributes output parameter.
+    /// </summary>
+    /// <param name="textColor">Specifies the text color.</param>
+    /// <returns>
+    /// Standard HRESULT error code.
+    /// </returns>
+    STDMETHOD(SetTextColor)(
+        DWRITE_COLOR_F const& textColor
+        ) PURE;
+
+    /// <summary>
+    /// Sets the current color palette index. The default value is zero. Changing the palette index can affect
+    /// the appearance of a glyph if its definition references colors in the color palette. If this is the case,
+    /// the SetCurrentGlyph method returns the DWRITE_PAINT_ATTRIBUTES_USES_PALETTE flag via the glyphAttributes
+    /// output parameter.
+    /// </summary>
+    /// <param name="textColor">Specifies the color palette index.</param>
+    /// <returns>
+    /// Standard HRESULT error code.
+    /// </returns>
+    STDMETHOD(SetColorPaletteIndex)(
+        UINT32 colorPaletteIndex
+        ) PURE;
+
+    /// <summary>
+    /// Sets a custom color palette with client-defined palette entries instead of using a font-defined color
+    /// palette. Changing the color palette can affect the appearance of a glyph if its definition references
+    /// colors in the color palette. If this is the case, the SetCurrentGlyph method returns the
+    /// DWRITE_PAINT_ATTRIBUTES_USES_PALETTE flag via the glyphAttributes output parameter.
+    /// </summary>
+    /// <param name="paletteEntries">Array of palette entries for the client-defined color palette.</param>
+    /// <param name="paletteEntryCount">Size of the paletteEntries array. This must equal the font's palette
+    /// entry count as returned by IDWriteFontFace2::GetPaletteEntryCount.</param>
+    /// <returns>
+    /// Standard HRESULT error code.
+    /// </returns>
+    STDMETHOD(SetCustomColorPalette)(
+        _In_reads_(paletteEntryCount) DWRITE_COLOR_F const* paletteEntries,
+        UINT32 paletteEntryCount
+        ) PURE;
+
+    /// <summary>
+    /// Sets the current position in the visual tree to the first child of the current paint element, and returns
+    /// the newly-selected element's properties via the paintElement output parameter.
+    /// </summary>
+    /// <param name="paintElement">Receives the properties of the newly-selected element.</param>
+    /// <param name="structSize">Size of the DWRITE_PAINT_ELEMENT structure, in bytes.</param>
+    /// <returns>
+    /// Standard HRESULT error code. The return value is E_INVALIDARG if the current paint element doesn't have
+    /// any children.
+    /// </returns>
+    /// <remarks>
+    /// Whether a paint element has children (and how many) can be determined a priori from its paint type and
+    /// properties. For more information, see DWRITE_PAINT_ELEMENT.
+    /// </remarks>
+    STDMETHOD(MoveToFirstChild)(
+        _Out_writes_bytes_(structSize) DWRITE_PAINT_ELEMENT* paintElement,
+        UINT32 structSize = sizeof(DWRITE_PAINT_ELEMENT)
+        ) PURE;
```

Comment:
I assume that if the `DWRITE_PAINT_ELEMENT` structure is extended in the future, the IDWritePaintReader will have to support different sizes corresponding to different versions of the API.

I also assume that all extensions of the `DWRITE_PAINT_ELEMENT` structure will consist of adding new members to the end, so that downlevel callers can just inspect the part they understand.

(One thing I'm not sure of is whether `IDWritePaintReader` is a DWrite-provided interface or an interface that consumers implement. I think it's DWrite-provided, so it will always support all of the things that DWrite does.)

Reply:
DISCUSSION: This field is written to by our implementation up to the size, and the caller has definitions of this struct based on the SDK and version level they are using.

RECOMMEND: Zero-fill the struct up to the provided size before use.

Reply:
Done in the implementation.